### PR TITLE
Allow users to specify initial_states for recurrent layers

### PR DESCRIPTION
**Pull request by lobachevzky**
Specifying the initial hidden state of a recurrent layer is necessary for sequence to sequence models, and probably for other architectures as well.

### CODE REVIEW

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -217,7 +218,15 @@ def call(self, x, mask=None):
...
                                 ': ' + str(input_shape))
-        if self.stateful:
+        if self.initial_states is not None:
+            if self.stateful:
+                raise Exception('Keras does not currently support stateful Recurrent '
+                                'layers with custom initial states.')
+            if type(self.initial_states) in (list, tuple):
+                initial_states = self.initial_states
+            else:
+                initial_states = [self.initial_states]
```
Comment (fchollet) [inline keras/layers/recurrent.py]: I don't get. You're aware this is supposed to be a symbolic variable, right?

Comment (braingineer) [inline keras/layers/recurrent.py]: I think this is meant to be like the `set_weights` function, but just for the initial hidden state (so you could have an RNN conditioned on something). 

@lobachevzky , you should model it more after: https://github.com/braingineer/keras/blob/master/keras/engine/topology.py#L831

Comment (lobachevzky) [inline keras/layers/recurrent.py]: What @braingineer said is correct but it's not a Numpy array. In the application where I've been using this, I use the output of another layer and it works fine. @fchollet, not sure what you mean by it being a symbolic variable.

Comment (fchollet) [inline keras/layers/recurrent.py]: I mean a symbolic tensor (e.g. TF or Theano tensor) as opposed to a numeric tensor (a Numpy array).

I don't think that most people would want to initialize the states via symbolic tensors. What is your use case?

Comment (lobachevzky) [inline keras/layers/recurrent.py]: Right. The use-case is [sequence to sequence models](https://papers.nips.cc/paper/5346-sequence-to-sequence-learning-with-neural-networks.pdf). In those models, the final hidden state of the first set of layers is the initial hidden state for the subsequent set of layers. There are some other features that are needed to fully implement the model described in that paper, but I'm currently working on them, so I should have another pull request ready soon.

**Comment by xingdi-eric-yuan** [comment]
@fchollet it would be very helpful if one can init the hidden states of rnn, there are a lot of NLP models (machine translation, text generation) use the sequence to sequence model, which is mentioned by @lobachevzky . Basically it's an encoder-decoder model, we need the hidden states of encoder to be the init states of decoder :)

**Comment by RamaneekGill** [comment]
Is this PR going to be fixed or will the development of the feature be continued in #3947 ?

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.

**Pull request closed without merging**